exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/handlers/CreateFunctionHandler.java

```diff
@@ -0,0 +1,195 @@
+              String.format("Jar with %s name is used. Action: %s", jarName, action));
+        }
+
+        // verify that binary and source exist
+        Path remoteBinary = new Path(remoteRegistry.getStagingArea(), jarName);
+        Path remoteSource = new Path(remoteRegistry.getStagingArea(), sourceName);
+        if (!fs.exists(remoteBinary) || !fs.exists(remoteSource)) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Binary [%s] or source [%s] is absent in udf staging area [%s].", jarName, sourceName, remoteRegistry.getStagingArea().toUri().getPath()));
+        }
+
+        // backup binary & source (copy to udf tmp directory)
+        fs.mkdirs(tmpDir);
+        Path tmpBinary = new Path(tmpDir, jarName);
+        Path tmpSource = new Path(tmpDir, sourceName);
+
+        FileUtil.copy(fs, remoteBinary, fs, tmpBinary, false, fs.getConf());
+        FileUtil.copy(fs, remoteSource, fs, tmpSource, false, fs.getConf());
+
+        // copy binary to local fs, we don't need source for validation
+        Path localBinary = new Path(new Path(localTmpDir.toURI()), jarName);
+        fs.copyToLocalFile(tmpBinary, localBinary);
+
+        // validate functions locally
+        List<Func> functions;
+        try {
+          functions = context.getFunctionRegistry().validate(localBinary);
+        } catch (FunctionValidationException ex) {
+          return DirectPlan.createDirectPlan(context, false, ex.getMessage());
+        }
+
+        if (functions.size() == 0) {
+          return DirectPlan.createDirectPlan(context, false,
+              String.format("Jar %s does not contain functions", jarName));
+        }
+
+        // validate and register remotely
+        Jar jar = Jar.newBuilder().setName(jarName).addAllFunction(functions).build();
+        String error = register(remoteRegistry, jar, tmpBinary, tmpSource, remoteRegistry.getRetryTimes());
+
+        if (error != null) {
+          return DirectPlan.createDirectPlan(context, false, error);
+        }
+
+        // remove jars from staging area
+        try {
+          fs.delete(remoteBinary, false);
+          fs.delete(remoteSource, false);
+        } catch (IOException ex) {
+          logger.warn("Failed to delete binary {} and source {} from staging area", jarName, sourceName);
+        }
+
+        return DirectPlan.createDirectPlan(context, true,
+            String.format("The following UDFs in jar %s have been registered:\n%s", jarName, functions));
+
+      } finally {
+        if (inProgress) {
+          remoteRegistry.removeFromJars(jarName);
+        }
+        FileUtils.deleteQuietly(localTmpDir);
+        if (fs.exists(tmpDir)) {
+          fs.delete(tmpDir, true);
+        }
+      }
+    }
+    throw UserException.validationError()
+        .message("Dynamic UDFs support is disabled.")
+        .build(logger);
+  }
+
+  private String validate(Registry registry, Jar jar) {
+    for (Jar remoteJar : registry.getJarList()) {
```

Comment:
Race condition. Like the above.

Drillbit X: reads (a, b, c)
Drillbit Y: reads (a, b, c)
Drillbit X writes (a, b, c, d)
Drllbit Y checks (a, b, c), sees it can add d and writes (a, b, c, d)

There needs to be a way to do the check and update attomically in some way.